### PR TITLE
Allow for passing stylesheet to `wp_theme_has_theme_json`.

### PR DESCRIPTION
**Pull request by spacedmonkey**
Part of https://github.com/WordPress/gutenberg/issues/45171

## What?
Allow stylesheet to `wp_theme_has_theme_json`. This allows for the developers to check if none active themes support theme json. This also improves caching, as caches are no longer depending on current theme. 

See https://github.com/WordPress/gutenberg/pull/45950 https://github.com/WordPress/gutenberg/pull/45831

### CODE REVIEW

File: lib/compat/wordpress-6.2/get-global-styles-and-settings.php [hunk 1]
```diff
@@ -25,23 +30,21 @@ function wp_theme_has_theme_json() {
...
 			return (bool) $theme_has_support;
 		}
 
-		// Has the own theme a theme.json?
-		$theme_has_support = is_readable( get_stylesheet_directory() . '/theme.json' ) ? 1 : 0;
-
-		// Look up the parent if the child does not have a theme.json.
-		if ( 0 === $theme_has_support ) {
-			$theme_has_support = is_readable( get_template_directory() . '/theme.json' ) ? 1 : 0;
+		$wp_theme = wp_get_theme( $stylesheet );
```
Comment (oandregal) [inline lib/compat/wordpress-6.2/get-global-styles-and-settings.php:40]: See conversation about performance of `wp_get_theme` at https://github.com/WordPress/gutenberg/pull/45831/files#r1029225791

File: lib/compat/wordpress-6.0/get-global-styles-and-settings.php [hunk 2]
```diff
@@ -85,7 +86,7 @@ function gutenberg_get_global_styles_svg_filters() {
 		}
 	}
 
-	$supports_theme_json = wp_theme_has_theme_json();
+	$supports_theme_json = wp_theme_has_theme_json( $stylesheet );
```
Comment (oandregal) [inline lib/compat/wordpress-6.0/get-global-styles-and-settings.php]: Do we need to do this? Isn't this code querying for the active theme (hence, no need for stylesheet)?

File: lib/compat/wordpress-6.1/get-global-styles-and-settings.php [hunk 3]
```diff
@@ -75,15 +76,15 @@ function gutenberg_get_global_stylesheet( $types = array() ) {
...
+	$transient_name = 'gutenberg_global_styles_' . $stylesheet;
 	if ( $can_use_cached ) {
 		$cached = get_transient( $transient_name );
 		if ( $cached ) {
 			return $cached;
 		}
 	}
 	$tree                = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data();
-	$supports_theme_json = wp_theme_has_theme_json();
+	$supports_theme_json = wp_theme_has_theme_json( $stylesheet );
```
Comment (oandregal) [inline lib/compat/wordpress-6.1/get-global-styles-and-settings.php]: Do we need to do this? Isn't this code querying the active theme (hence, no need for stylesheet)?

File: lib/compat/wordpress-6.2/get-global-styles-and-settings.php [hunk 4]
```diff
@@ -12,11 +12,16 @@
...
 	 */
-	function wp_theme_has_theme_json() {
-		$cache_group       = 'theme_json';
-		$cache_key         = 'wp_theme_has_theme_json';
+	function wp_theme_has_theme_json( $stylesheet = '' ) {
+		$cache_group = 'theme_json';
+		if ( empty( $stylesheet ) ) {
+			$stylesheet = get_stylesheet();
+		}
+		$cache_key         = sprintf( 'wp_theme_has_theme_json_%s', $stylesheet );
```
Comment (oandregal) [inline lib/compat/wordpress-6.2/get-global-styles-and-settings.php:24]: I see two issues with caching by theme:

- Code complexity: what are the benefits of having cache keys per theme when we only need one at a time?
- I don't see a way to clean the cache key for the previous theme upon a `start_previewing_event`. It gives the stylesheet of the theme being previewed, but at that point we need to clear the cache key of the _previous_ theme, not the one being previewed. How can we do this?

File: lib/compat/wordpress-6.2/get-global-styles-and-settings.php [hunk 5]
```diff
@@ -51,9 +54,40 @@ function wp_theme_has_theme_json() {
...
+if ( ! function_exists( '_wp_theme_has_theme_json_clean_cache_start_previewing_theme' ) ) {
+	/**
+	 * Clear on preview.
+	 *
+	 * @param WP_Customize_Manager $manager WP_Customize_Manager instance.
 	 */
-	function wp_theme_has_theme_json_clean_cache() {
-		wp_cache_delete( 'wp_theme_has_theme_json', 'theme_json' );
+	function _wp_theme_has_theme_json_clean_cache_start_previewing_theme( $manager ){
+		wp_theme_has_theme_json_clean_cache( $manager->get_stylesheet() );
```
Comment (oandregal) [inline lib/compat/wordpress-6.2/get-global-styles-and-settings.php]: See https://github.com/WordPress/gutenberg/pull/45955/files#r1029244275 we need to clear the cache of the _previous_ theme, not the actual one.

File: lib/compat/wordpress-6.2/get-global-styles-and-settings.php [hunk 6]
```diff
@@ -51,9 +54,40 @@ function wp_theme_has_theme_json() {
...
+if ( ! function_exists( '_wp_theme_has_theme_json_clean_cache_switch_theme' ) ) {
+	/**
+	 * Clean new and old themes on switch.
+	 *
+	 * @param string  $new_name Name of the new theme.
+	 * @param WP_Theme $new_theme WP_Theme instance of the new theme.
+	 * @param WP_Theme $old_theme WP_Theme instance of the old theme.
+	 */
+	function _wp_theme_has_theme_json_clean_cache_switch_theme( $new_name, $new_theme, $old_theme ) {
+		wp_theme_has_theme_json_clean_cache( $new_theme->get_stylesheet() );
```
Comment (oandregal) [inline lib/compat/wordpress-6.2/get-global-styles-and-settings.php]: See https://github.com/WordPress/gutenberg/pull/45955/files#r1029244275

File: lib/compat/wordpress-6.0/get-global-styles-and-settings.php (hunk 2 above)
Comment (spacedmonkey) [reply lib/compat/wordpress-6.0/get-global-styles-and-settings.php]: This saves a another lookup to `get_option`. We have the stylesheet to hand, so why not just pass it in and save a lookup to cache.

File: lib/compat/wordpress-6.1/get-global-styles-and-settings.php (hunk 3 above)
Comment (spacedmonkey) [reply lib/compat/wordpress-6.1/get-global-styles-and-settings.php]: https://github.com/WordPress/gutenberg/pull/45955/files#r1029270566

File: lib/compat/wordpress-6.2/get-global-styles-and-settings.php (hunk 1 above)
Comment (spacedmonkey) [reply lib/compat/wordpress-6.2/get-global-styles-and-settings.php:40]: This allows us to get if another theme supports theme json. It is needed for [here](https://github.com/WordPress/wordpress-develop/blob/239a6c33ec5fef9416655c36225b064b987c31cc/src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php#L372) and [here](https://github.com/WordPress/wordpress-develop/blob/239a6c33ec5fef9416655c36225b064b987c31cc/src/wp-includes/class-wp-theme-json-resolver.php#L423-L425).

File: lib/compat/wordpress-6.2/get-global-styles-and-settings.php (hunk 4 above)
Comment (spacedmonkey) [reply lib/compat/wordpress-6.2/get-global-styles-and-settings.php:24]: Complexity is not really a good argument not to do something. Just becuase it is hard, does not mean it not worthwhile. I have done the hard work to make this work, let's go ahead here. 

Good point about `start_previewing_event`. This cache clearing can be removed, as now the cache key will be different once you start previewing, which is pretty neat. I am not against removing this all together, but I kept, just in case.

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php [hunk 7]
```diff
@@ -16,6 +16,81 @@
...
+	public static function get_user_data_from_wp_global_styles( $theme, $create_post = false, $post_status_filter = array( 'publish' ) ) {
+		if ( ! $theme instanceof WP_Theme ) {
+			$theme = wp_get_theme();
+		}
+
+		/*
+		 * Bail early if the theme does not support a theme.json.
+		 *
+		 */
+		if ( ! wp_theme_has_theme_json( $theme->get_stylesheet() ) ) {
```
Comment (oandregal) [inline lib/experimental/class-wp-theme-json-resolver-gutenberg.php]: I think these inner private methods should not hold this kind of domain logic. It should be the consumer responsibility to ask for the data they need. This is how I envision improving on the current situation:

- 1st step is making the `get_merged_data` method work properly (take into account the origin) https://github.com/WordPress/gutenberg/pull/45969
- 2nd step is modifying the public methods to ask for the proper origin of data they want. Example for the settings https://github.com/WordPress/gutenberg/pull/45971

File: lib/compat/wordpress-6.2/get-global-styles-and-settings.php (hunk 4 above)
Comment (oandregal) [reply lib/compat/wordpress-6.2/get-global-styles-and-settings.php:24]: > Complexity is not really a good argument not to do something. Just becuase it is hard, does not mean it not worthwhile. I have done the hard work to make this work, let's go ahead here.

My point is not about doing something that is hard. It is about not doing something complex when it can be simple.

This approach is more complex in terms of code and more demanding in terms of memory (we store cache keys we don't need instead of one). The direction I shared a while ago and summarized [here](https://github.com/WordPress/gutenberg/pull/45955/files#r1029306186) again is simpler and brings bigger performance gains. Would you be willing to give it a try and reevaluate when is it shipped?

Comment (spacedmonkey) [reply lib/compat/wordpress-6.2/get-global-styles-and-settings.php:24]: I don't understand this feedback at all. For most requests, it will only store one cache key, like it does currently. There difference here, is this structure all for multiple themes theme json to be required. This makes cache invalidation easier, not harder. 

Where is the there a performance lose here?

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php (hunk 7 above)
Comment (spacedmonkey) [reply lib/experimental/class-wp-theme-json-resolver-gutenberg.php]: On this we disagree. The current system of origin, is hard to understand, hard to test and VERY unclaer from a development standpoint. We should be moving away from this pattern and the code easier to use any understand.

Comment (oandregal) [reply lib/experimental/class-wp-theme-json-resolver-gutenberg.php]: The fundamental issue with adding caching to the resolver instead of adding it to the public methods is this: every public method has different conditions for cache invalidation. settings & styles need some, theme.json check needs some other, and patterns/custom templates need totally different conditions.

We cannot simply replicate those invalidations within the resolver. That was the core issue we ran into in the WordPress 6.1 cycle.

Comment (spacedmonkey) [reply lib/experimental/class-wp-theme-json-resolver-gutenberg.php]: This method `get_user_data_from_wp_global_styles` take a param of `WP_Theme`, we need to respect that and check the requested them support theme.json. 

Look at this method in core. https://github.com/WordPress/wordpress-develop/blob/9ed27339b19eb07aa01bcb7e593be2e11f928f1e/src/wp-includes/class-wp-theme-json-resolver.php#L423-L425

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php [hunk 8]
```diff
@@ -68,7 +141,7 @@ public static function get_theme_data( $deprecated = array(), $settings = array(
 					 */
 					$parent_theme->merge( static::$theme );
 					static::$theme = $parent_theme;
-				}
+				}
```
Comment (spacedmonkey) [inline lib/experimental/class-wp-theme-json-resolver-gutenberg.php]:
```suggestion
				}
```

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php [hunk 9]
```diff
@@ -56,9 +130,8 @@ public static function get_theme_data( $deprecated = array(), $settings = array(
 
 			if ( $wp_theme->parent() ) {
 				// Get parent theme.json.
-				$parent_theme_json_file = static::get_file_path_from_theme( 'theme.json', true );
-				if ( '' !== $parent_theme_json_file ) {
-					$parent_theme_json_data = static::read_json_file( $parent_theme_json_file );
+				if ( wp_theme_has_theme_json( $wp_theme->parent()->get_stylesheet() ) ) {
```
Comment (spacedmonkey) [inline lib/experimental/class-wp-theme-json-resolver-gutenberg.php]: @felixarntz @oandregal Is the perfect example of why we need this function to allow for passing a stylesheet. In this context, we need to know if the parent ( not the current ) theme has theme json. This is need for every request, making this change required.

File: lib/compat/wordpress-6.2/get-global-styles-and-settings.php (hunk 1 above)
Comment (felixarntz) [reply lib/compat/wordpress-6.2/get-global-styles-and-settings.php:40]: To be honest, I think this conversation about performance of `wp_get_theme` has blown out of proportion a little. Yes, if we can avoid calling it 5 times in a single function, we should rather call it once and put it in a variable.

However, that is a micro optimization that I don't think is critical. I haven't seen any performance testing that would show us it is truly a concern; of course instantiating an object over and over is more expensive than not instantiating the object again, but to me this seems more like a nit-pick concern that won't matter for actual performance.

In other words, it is fine to introduce a call to `wp_get_theme`. Just don't call it when you don't have to. In any case, I think here we don't have to (see my overarching feedback).

**Review by felixarntz — COMMENTED** [review]
@oandregal @spacedmonkey Chiming in here late, I see both of your concerns:

1. I agree with you @spacedmonkey that we should provide a way to find out whether a theme has `theme.json` for _any_ theme, not just the current theme.
2. I also agree with you @oandregal that changing this `wp_theme_has_theme_json()` function to support any theme should not be necessary.

The main use-case surely is the current theme. However, WordPress has always provided ways to get the same information for any theme, so we should work on that too. But my main feedback is that those things should happen in different places to align with how WordPress core works today.

Consider the following: WordPress core today has theme _functions_, those typically work only for the current theme. But it also has the `WP_Theme` class that provides _methods_ to get the same data for any theme (whatever you instantiate the `WP_Theme` with, via `wp_get_theme()`).

Some examples:
* `get_stylesheet()` vs `WP_Theme::get_stylesheet()`
* `get_template()` vs `WP_Theme::get_template()`
* `get_template_directory()` vs `WP_Theme::get_template_directory()`
* `get_theme_file_path()` vs `WP_Theme::get_file_path()`

In all of the above, the function is always for the current theme (no parameter), and the method works for whatever `WP_Theme` is instantiated as. In other words, I would propose we follow the same approach:
* Keep `wp_theme_has_theme_json()` simple, without a parameter and without having to use `wp_get_theme()`.
* Introduce a method `WP_Theme::has_theme_json()` that allows doing the same for any theme.

That's my 2 cents.

File: lib/compat/wordpress-6.2/get-global-styles-and-settings.php (hunk 1 above)
Comment (spacedmonkey) [reply lib/compat/wordpress-6.2/get-global-styles-and-settings.php:40]: @felixarntz YES! In this case, we call it once and after that, we cache. This is called once per page load. Making it fine. Calling it 5k times per page is the problem.

File: lib/experimental/class-wp-theme-json-resolver-gutenberg.php (hunk 7 above)
Comment (spacedmonkey) [reply lib/experimental/class-wp-theme-json-resolver-gutenberg.php]: @felixarntz add this change. Can you chime in here.

**Comment by carolinan** [comment]
> check if none active themes support theme json.

I believe we will need this change for https://github.com/WordPress/gutenberg/issues/45371 which is part of Phase 2: Customization

**Comment by felixarntz** [comment]
@spacedmonkey Per my comment in https://github.com/WordPress/gutenberg/pull/45955#pullrequestreview-1220128114, I still don't think this is a good idea. IMO we should rather introduce a `has_theme_json()` method to `WP_Theme` to support that use-case. Then you can call `wp_get_theme( ... )->has_theme_json()`.
